internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,240 @@
+	cacheGvks := c.getCacheGVKs()
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	type startResult struct {
+		source CloserSyncingSource
+		gvk    schema.GroupVersionKind
+		err    error
+	}
+	startResults := make(chan startResult)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			source, err := c.startNewSource(ctx, gvk, watcher)
+			startResults <- startResult{
+				source: source,
+				gvk:    gvk,
+				err:    err,
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(startResults)
+	}()
+
+	sourcesErrors := []error{}
+	for result := range startResults {
+		if result.err != nil {
+			sourcesErrors = append(sourcesErrors, result.err)
+			continue
+		}
+
+		err := c.addSource(result.gvk, result.source)
+		if err != nil {
+			// If we made it here then there is a logic error in
+			// calculating the diffs between what is currently being
+			// watched by the cache
+			panic(err)
+		}
+	}
+
+	return errors.Join(sourcesErrors...)
+}
+
+func (c *cache) startNewSource(ctx context.Context, gvk schema.GroupVersionKind, watcher Watcher) (CloserSyncingSource, error) {
+	s, err := c.sourcerer.Source(gvk, c.owner, func(ctx context.Context) {
+		err := c.removeSource(gvk)
+		if err != nil {
+			logr := log.FromContext(ctx)
+			logr.Error(err, "managed content cache postSyncError removing source failed", "gvk", gvk)
+		}
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting source: %w", err)
+	}
+
+	err = watcher.Watch(s)
+	if err != nil {
+		return nil, fmt.Errorf("establishing watch for GVK %q: %w", gvk, err)
+	}
+
+	syncCtx, syncCancel := context.WithTimeout(ctx, c.syncTimeout)
+	defer syncCancel()
+	err = s.WaitForSync(syncCtx)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for sync: %w", err)
+	}
+
+	return s, nil
+}
+
+func (c *cache) addSource(gvk schema.GroupVersionKind, source CloserSyncingSource) error {
+	if _, ok := c.sources[gvk]; !ok {
+		c.sources[gvk] = source
+		return nil
+	}
+	return errors.New("source already exists")
+}
+
+func (c *cache) removeStaleSources(gvks sets.Set[schema.GroupVersionKind]) error {
+	cacheGvks := c.getCacheGVKs()
+	removeErrs := []error{}
+	gvksToRemove := cacheGvks.Difference(gvks)
+	for _, gvk := range gvksToRemove.UnsortedList() {
+		err := c.removeSource(gvk)
+		if err != nil {
+			removeErrs = append(removeErrs, err)
+		}
+	}
+	return errors.Join(removeErrs...)
+}
+
+func (c *cache) removeSource(gvk schema.GroupVersionKind) error {
+	if source, ok := c.sources[gvk]; ok {
+		err := source.Close()
+		if err != nil {
+			return fmt.Errorf("closing source for GVK %q: %w", gvk, err)
+		}
+	}
+	delete(c.sources, gvk)
+	return nil
+}
+
+func (c *cache) getCacheGVKs() sets.Set[schema.GroupVersionKind] {
+	cacheGvks := sets.New[schema.GroupVersionKind]()
+	for gvk := range c.sources {
+		cacheGvks.Insert(gvk)
+	}
+	return cacheGvks
```

Comment:
Just an FYI: 1.23 introduces `maps.Values` in the stdlib as an iterator, which I suspect would slightly simplify this function. Not sure where we are on bumping to 1.23, but thought I'd mention this.

Reply:
nice!